rfcs/convergence/remove-commons.md

```diff
@@ -0,0 +1,140 @@
+
+```ts
+export type CheckboxCommons = {
+  // These are required in state (made optional in props)
+  size: 'medium' | 'large';
+};
+
+export type CheckboxProps =
+  // Things in CheckboxCommons are optional in props
+  Partial<CheckboxCommons> &
+    ComponentProps<CheckboxSlots> & {
+      // maybe some other props here
+    };
+
+export type CheckboxState =
+  // Commons are required in state
+  CheckboxCommons &
+    ComponentState<CheckboxSlots> & {
+      // maybe some other state here
+    };
+```
+
+## Problem statement
+
+Our current use of `FooCommons` interfaces as part of component props could cause unnecessary confusion for a user looking at our types:
+
+- Not straightforward to understand what's optional or required when
+- Additional level of indirection to understand what's included in props
+- It's an implementation detail that the user shouldn't need to care about when looking at props typings
+
+We're also inconsistent about how required/optional things are handled in commons and props--whether things are:
+
+- optional in commons type (to reflect props) then do `type FooState = Required<FooCommons> & ...`.
+- required in commons type (to reflect state) then do `type FooProps = Partial<FooCommons> & ...`
+
+## Detailed Design or Proposal
+
+Rather than making a consistent rule for whether commons should more closely reflect props or state, I think a better solution is to remove the `FooCommons` types entirely. Instead, put everything in `FooProps`, and use `Pick`/`Required` to define `FooState` (and `FooContext` if needed).
+
+This makes the user-facing type (props) very clear/explicit and confines the implementation details to the internal types (state, context).
+
+Modified Input example from above:
+
+```ts
+export type InputProps = ComponentProps<InputSlots> & {
+  size?: 'small' | 'medium' | 'large';
+  appearance?: 'outline' | 'underline' | 'filledDarker' | 'filledLighter';
+  // maybe some other props here
+};
+
+export type InputState =
+  // If there are things from props which remain optional in state, add another Pick without Required
+  Required<Pick<InputProps, 'size' | 'appearance'>> &
+    ComponentState<InputSlots> & {
+      // maybe some other state here
+    };
+```
+
+Modified Checkbox example (now using an identical pattern):
+
+```ts
+export type CheckboxProps = ComponentProps<CheckboxSlots> & {
+  size?: 'medium' | 'large';
+  // maybe some other props here
+};
+
+export type CheckboxState = Required<Pick<CheckboxProps, 'size'>> &
+  ComponentState<CheckboxSlots> & {
+    // maybe some other state here
+  };
+```
+
+As an interim step (to avoid blocking RC), we could **stop exporting `FooCommons` types**. This way, if the actual changes aren't finished before RC, we could finish changing the types later without breaking public APIs (since the final included props and their optional/required status should be the same either way).
+
+### Pros
+
+- User-facing types (props) will be very clear/explicit
```

Comment:
hmm lets say if one needs all props in State as well, then it's gonna be even worse, not sure how much this situation can/is occurring in our codebase.

Reply:
state types will be user facing, if user want's to implement some of functionality on its own AFAIR. While in one scenario this will improve things, it will make the code more cluttered  for maintainers/contributors and then for consumers as well ( if they use only some parts of the component api )

Reply:
The state types are technically user-facing, but only a small percentage of users will need to care (and those users will probably be the ones who are more willing/competent to dig in and understand things). Hence the suggestion to optimize for clarity of props rather than state.

Part of the reason I'm advocating to optimize for clarity in the most common case is based on my experience with maintaining a widely-used component library: some users can be pretty easily confused about what props are available, and it's worse the more levels of indirection are added (I'm concerned about this with the ComponentProps type, but that's a separate discussion).